src/infer_request.h

```diff
@@ -680,6 +679,20 @@ class InferenceRequest {
     secondary_stats_aggregator_ = secondary_stats_aggregator;
   }
 
+  Status Cancel()
+  {
+    if (state_ != InferenceRequest::State::INITIALIZED) {
+      response_factory_->Cancel();
+    } else {
+      return Status(
+          Status::Code::INTERNAL,
+          "Request cannot be cancelled before it has started executing.");
```

Comment:
We may want to allow cancellation before Run() is called to make frontend life simpler -

But things to consider: is there any race condition between initialized and cancel? Cancel could be called from any thread as it is public. (I don't think there would be - but not sure) 

Also would tweak the message - its not necessarily executing - but before passed to InferAsync .. not sure best way to rephrase that though .


```suggestion
          "Request cannot be cancelled before it has been passed to Infer_Async");
```

Reply:
> We may want to allow cancellation before Run() is called to make frontend life simpler -

I don't think it would make the frontend life simpler. Frontend needs to make sure to only cancel the request after TRITONSERVER_InferAsync. This would be the same even if we allow cancelling before Run is called.

> But things to consider: is there any race condition between initialized and cancel? Cancel could be called from any thread as it is public. (I don't think there would be - but not sure)

I converted the request state to atomic. As long as cancel is called after initialize, I think we should be fine.

Will update the message.

Reply:
One thing - we still need to reset cancelled status I believe on Run/Prepare - I may have missed it

Reply:
Addressed the comment.